File: Computer Programming/C++/Recursion and Dynamic Programming/Unique Paths.cpp
class Solution {
public:
    int uniquePaths(int m, int n) 
    {
        vector<vector<int>>vis(m,vector<int>(n,-1));

        int ans =  paths(m-1,n-1,vis);      
        return ans;
    }
    int paths(int i,int j, vector<vector<int>> &vis)
    {
        if(i<0 || j<0)
        {
            return 0;
        }
        if(i==0 && j==0)
        {
            return 1;
        }
        if(vis[i][j] != -1)
        {
            return vis[i][j];
        }

        int top = paths(i-1,j,vis);
        int left = paths(i,j-1,vis);
        vis[i][j] = top + left;
        return vis[i][j];
    }
};

File: Computer Programming/Python/Dynamic Programming/Unique Paths.py
class Solution(object):
    def uniquePaths(self, m, n):
        """
        :type m: int
        :type n: int
        :rtype: int
        """
        
        dp = []
        for i in range(0,m):
            curr = []
            for j in range(0,n):
                curr.append(-1)
            dp.append(curr)


        return self.function(dp,m-1,n-1)

    def function(self,dp,row,col):
        if row == 0 and col == 0:
            return 1
        if row < 0 or col < 0:
            return 0
    
        if dp[row][col] != -1:
            return dp[row][col]

        left = self.function(dp,row,col-1)
        down = self.function(dp,row-1,col)

        dp[row][col] = left + down
        return dp[row][col]
